Sources/Document/CRDT/RGATreeList.swift

```diff
@@ -0,0 +1,422 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import Foundation
+
+/**
+ * `RGATreeListNode` is a node of RGATreeList.
+ */
+class RGATreeListNode: SplayNode<CRDTElement> {
+    private var previous: RGATreeListNode?
+    private var next: RGATreeListNode?
+
+    override init(_ value: CRDTElement) {
+        super.init(value)
+    }
+
+    /**
+     * `create` creates a new node after the previous node.
+     */
+    static func create(with value: CRDTElement, previousNode: RGATreeListNode) -> RGATreeListNode {
+        let newNode = RGATreeListNode(value)
+        let prevNext = previousNode.next
+        previousNode.next = newNode
+        newNode.previous = previousNode
+        newNode.next = prevNext
+        if let prevNext = prevNext {
+            prevNext.previous = newNode
+        }
+
+        return newNode
+    }
+
+    /**
+     * `getLength` returns the length of this node.
+     */
+    override func getLength() -> Int {
+        return self.value.isRemoved() ? 0 : 1
+    }
+
+    /**
+     * `remove` removes value based on removing time.
+     */
+    func remove(_ removedAt: TimeTicket) -> Bool {
+        return self.value.remove(removedAt)
+    }
+
+    /**
+     * `getCreatedAt` returns creation time of this value
+     */
+    func getCreatedAt() -> TimeTicket {
+        return self.value.getCreatedAt()
+    }
+
+    /**
+     * `getPositionedAt` returns time this element was positioned in the array.
+     */
+    func getPositionedAt() -> TimeTicket {
+        if let movedAt = self.value.getMovedAt() {
+            return movedAt
+        }
+
+        return self.value.getCreatedAt()
+    }
+
+    /**
+     * `release` releases prev and next node.
+     */
+    func release() {
+        if let previous = self.previous {
+            previous.next = self.next
+        }
+        if let next = self.next {
+            next.previous = self.previous
+        }
+        self.previous = nil
+        self.next = nil
+    }
+
+    /**
+     * `getPrev` returns a previous node.
+     */
+    func getPrevious() -> RGATreeListNode? {
+        return self.previous
+    }
+
+    /**
+     * `getNext` returns a next node.
+     */
+    func getNext() -> RGATreeListNode? {
+        return self.next
+    }
+
+    /**
+     * `isRemoved` checks if the value was removed.
+     */
+    @discardableResult
+    func isRemoved() -> Bool {
+        return self.value.isRemoved()
+    }
+}
+
+/**
+ * `RGATreeList` is replicated growable array.
+ */
+class RGATreeList: Sequence, IteratorProtocol {
+    private let dummyHead: RGATreeListNode
+    private var last: RGATreeListNode
+    private var nodeMapByIndex: SplayTree<CRDTElement>
+    private var nodeMapByCreatedAt: [String: RGATreeListNode]
```

Comment:
Can we use TimeTicket instead of String? 🤔

```suggestion
    private var nodeMapByCreatedAt: [TimeTicket: RGATreeListNode]
```

Reply:
thank you for your good suggestion.

Reply:
I have changed TimeTicket's type from class to struct and applied Hashable and CustomConvertableString to TimeTicket.